### DISCUSSION TITLE
Async computed/derived prop without suspense or loadable, keep previous value

### DISCUSSION THREAD

**Discussion by vincerubinetti** [pmndrs/jotai#2004] [Q&A] [answered]
I have the following (simplified) situation:

I have an `text` atom that gets set to the content of a user-uploaded text file. Then I have some `computed` state from that, including some properties that have to be async.

```ts
const text = atom("");

const computed = atom(
  async (get) => {
    const asyncProp = await someAsyncFunc(get(text));
    return { asyncProp };
  }
);
```

This works fine, but now that atom is a promise. By default, `useAtom` uses React suspense and error boundaries for these types of atoms, but I can't use those. The reason is, the text content from the file the user uploads also gets put into an editable `<textarea>`, AND some of the computed properties are used on that same element. If I put a suspense boundary, no matter how low in the component tree, it will cause the textarea to disappear momentarily while `someAsyncFunc` runs and suspense renders the fallback, causing an ugly flicker and also losing the user's focus and cursor position in the textarea which is a pain to work around.

What I really want, I guess, is for the previous value of the computed atom to stay in place while the async functions run.

I saw in the docs that I can use the [loadable api](https://jotai.org/docs/utilities/async#loadable):

```tsx
const Component = () => {
  const [value] = useAtom(computed);

  // access computed value
  console.log(value.data?.asyncProp)
}
```

It actually seems to sort of keep the previous value of `data` while the next one is loading, leaving my UI usable. However, I'm getting this TypeScript error if I don't type-narrow for `state === "hasData"` (which I can't do, because I can't conditionally render my textarea, per above): 

```
Property 'data' does not exist on type 'Loadable<Promise<{ blah blah blah }>>'.
  Property 'data' does not exist on type '{ state: "loading"; }'.
```

Is this an error with the types? Is it supposed to keep the previous state? What am I missing here?

Would Valtio be good here? I saw that you can't do async in watchers at the moment.

**Comment by vincerubinetti** [comment]
> It actually seems to sort of keep the previous value of data while the next one is loading,

> Is this an error with the types?

It seems I was mistaken on this, it's not keeping the previous state (I was seeing the previous value due to an unrelated bug in my code).

So for more clarity, my component looks sorta like this (again very simplified):

```tsx
const [getText, setText] = useAtom(text);
const [getComputed] = useAtom(computed);

<textarea
  value={getText}
  onChange={(event) => setText(event.target.value)}
  tooltip={getComputed.data.asyncProp}
  // ts fix
  tooltip={getComputed.state === "hasData" && getComputed.data.asyncProp}
/>
```

In this particular case, I can get away with having the previous computed value disappear temporarily, but I'd still like to know if it's possible to create a recipe for a computed atom that just updates whenever the async computing finishes, where I don't want loading or error states to be raised to React suspense or error boundaries. Or perhaps `loadable` could be modified to have an option to keep previous data? I believe tanstack query has an option like this, `keepPreviousData`.

I'm looking at the `loadable` source code to see if I can figure it out. Seems like if I make sure it doesn't return a promise or throw an error, it should work as expected. Guidance or suggestions would be helpful. I also don't care about loading/error states in this particular case, so I might want a custom higher-order atom that makes it more convenient, so I don't have to do `.data` everywhere.

**Comment by dai-shi** [comment, marked as answer]
We have `unstable_unwrap` util that will be stable soonish.
```js
const syncAtom = unwrap(asyncAtom, (prev) => prev ?? "default value")
```

I had thought about extending `loadable` which does pretty similar, but their use cases are slightly different, so `unstable_unwrap` is introduced in jotai v2.
